pkg/algorithm/cell_allocation.go

```diff
@@ -89,21 +90,59 @@ func getLowestFreeCellLevel(freeList ChainCellList, l CellLevel) CellLevel {
 		"even split to the highest level %v", l-1))
 }
 
+// when buddyAlloc allocates bad cells,
+// check whether it is safe to split a higher level cell to get current level cells.
+func checkSplitSafety(freeList ChainCellList, freeCellNum map[CellLevel]int32, l CellLevel) bool {
+	var splitableCell Cell
+	splitableNum := map[CellLevel]int32{}
+	for i := CellLevel(len(freeList)); i >= CellLevel(1); i-- {
+		// calculate splitable number
+		splitableNum[i] = int32(len(freeList[i])) - freeCellNum[i]
+		if i < CellLevel(len(freeList)) && splitableCell != nil {
+			splitableNum[i] += splitableNum[i+1] * int32(len(splitableCell.GetChildren()))
+		}
+		// iterate higher level cell
+		if splitableCell == nil && len(freeList[i]) > 0 {
+			splitableCell = freeList[i][0]
+		} else if splitableCell != nil {
+			splitableCell = splitableCell.GetChildren()[0]
+		}
+		// check safety
+		if splitableNum[i] < 0 {
+			return false
+		}
```

Comment:
When will splitableNum[i] < 0? 
Seems we already safety break, and the whole scheduler should crash.

Reply:
in a test case, will change to panic later

Reply:
updated

Reply:
print more debug info when VC Safety Broken

---
In reply to: [451561941](https://github.com/microsoft/hivedscheduler/pull/27#discussion_r451561941) [](ancestors = 451561941)

Reply:
added